torch/lib/c10d/TCPStore.cpp

```diff
@@ -591,6 +591,229 @@ void TCPStoreMasterDaemon::run() {
 }
 #endif
 
+namespace detail {
+namespace {
+
+// Offers RAII for TCP sockets.
+class TCPSocket {
+ public:
+  TCPSocket() noexcept = default;
+
+  /* implicit */ TCPSocket(int handle) noexcept : handle_{handle} {}
+
+  TCPSocket(const TCPSocket& other) = delete;
+
+  TCPSocket& operator=(const TCPSocket& other) = delete;
+
+  TCPSocket(TCPSocket&& other) noexcept : handle_{other.handle_} {
+    other.handle_ = c10::nullopt;
+  }
+
+  TCPSocket& operator=(TCPSocket&& other) noexcept {
+    closeSocket();
+
+    handle_ = std::exchange(other.handle_, c10::nullopt);
+
+    return *this;
+  }
+
+  ~TCPSocket() {
+    closeSocket();
+  }
+
+  int handle() const noexcept {
+    return handle_.value_or(-1);
+  }
+
+ private:
+  void closeSocket() noexcept {
+    if (handle_) {
+      tcputil::closeSocket(*handle_);
+    }
+  }
+
+  c10::optional<int> handle_{};
+};
+
+} // namespace
+
+// Manages the lifecycle of a server daemon.
+class TCPServer {
+ public:
+  static std::shared_ptr<TCPServer> start(const TCPStoreOptions& opts);
+
+  PortType port() const noexcept {
+    return port_;
+  }
+
+  explicit TCPServer(
+      TCPSocket&& socket,
+      PortType port,
+      std::unique_ptr<TCPStoreMasterDaemon>&& daemon)
+      : socket_{std::move(socket)}, port_{port}, daemon_{std::move(daemon)} {}
+
+ private:
+  TCPSocket socket_;
+  PortType port_;
+  std::unique_ptr<TCPStoreMasterDaemon> daemon_;
+};
+
+std::shared_ptr<TCPServer> TCPServer::start(const TCPStoreOptions& opts) {
+  if (opts.multiTenant) {
+    LOG(WARNING)
+        << "The multi-tenant feature of TCPStore is not implemented yet.";
+  }
+
+  TCPSocket socket{};
+  PortType port{};
+
+  std::tie(socket, port) = tcputil::listen(opts.port);
+
+  auto daemon = std::make_unique<TCPStoreMasterDaemon>(socket.handle());
+
+  return std::make_shared<TCPServer>(
+      std::move(socket), port, std::move(daemon));
```

Comment:
Previously, TCPStoreMasterDaemon was responsible for shutting down and cleaning up the socket that it used once it was done (see `BackgroundThread::~BackgroundThread()`). Now that we have this new TCPSocket class where each socket can clean itself up, I think we should also remove this part of the code:
https://github.com/pytorch/pytorch/blob/ec89bf65352e0cf1e24738cd383c94785db3d4c2/torch/lib/c10d/TCPStore.cpp#L50-L54